### DISCUSSION TITLE
RFC - consistent GraphQL API Schema between roles

### DISCUSSION THREAD

**Discussion by ryami333** [directus/directus#21805] [Feature Requests]
AKA - "don't use the GraphQL schema as an access-control mechanism."

### Summary

One of the fundamental reasons why GraphQL is so advantageous as a front-end application query language is that the GraphQL schema acts as a universal (and **machine-readable**) _contract_ for the shape of data that crosses the application's data boundaries. In other words, its value is that we can predict the _exact_ shape of a query's response payload, by cross-referencing only the query itself and the GraphQL schema, and _nothing else_.

However, Directus has subverted this expectation by implementing dynamic GraphQL schema. Two different users will have different _schema_, depending on their access permissions to each of the available collections. This means we can no longer effectively use ubiquitous tools like [GraphQL Codegen](https://the-guild.dev/graphql/codegen), because the generated types only reflect one role's GraphQL schema. Two queries that are otherwise identical (besides headers) can be validated against entirely different schema. This problem is likely to be exacerbated even further once [policies](https://github.com/directus/directus/milestone/177) are implemented.

We are requesting that the GraphQL API behave more like other headless CMS solutions' (eg. Contentful, Sanity.io et al.) and decouple the GraphQL schema from collections' access-control, so that GraphQL queries can be statically analysed. We recognise that this would be a breaking change, and therefore would need to be tied to a major Directus release (eg. v11).

### Motivation

We want a single source-of-truth for our GraphQL APIs schema - and access-control to be enforced only at the server level (which it is already). This makes it possible to statically analyse `.graphql` files, which enables eg.
- Linting and (statically) validating queries.
- Using code-generators like [GraphQL Codegen](https://the-guild.dev/graphql/codegen).

### Detailed Design

All outlined above.

### Requirements List

All outlined above.

### Drawbacks

This would be a breaking change to _tooling+integrations_, which would necessitate a major Directus release (eg. v11.0.0).

However, it's worth pointing out that this would be a non-breaking change to _queries_. In other words, the runtime behaviour is more-or-less unchanged. A query that worked before will work now, and a query that failed before will fail now, except it might fail with `403 FORBIDDEN` instead of `400 FAILED VALIDATION`.

### Alternatives

Alternative is to do nothing.

### Adoption Strategy

See drawbacks, above.

**Comment by rijkvanzanten** [comment]
> One of the fundamental reasons why GraphQL is so advantageous as a front-end application query language is that the GraphQL schema acts as a universal (and machine-readable) contract for the shape of data that crosses the application's data boundaries

The contract and shape of the data differs based on the access control though. For example access to nested relational data changes the shape of the output, hence the contract. If you were to use a shared contract for every user, you'll have to include the _whole_ schema, which will not be accurate for users that have access control (and therefore exposing the shape of your data model to non-authenticated users if you have introspection enabled).

---

Now riffing on that thought: wouldn't you _want_ your front-end GraphQL contract to be based on the actual access you have from that front-end paradigm? If have a GraphQL file that documents the whole data model, but your user only has access to some of it, wouldn't you find yourself constantly running into unexpected unauthorized errors as the graphql tooling allowed you to make the requests without warning? 🤔

**Reply by rijkvanzanten** [reply]
I guess the TL;DR is:

> [...] This makes it possible to statically analyse .graphql files, which enables eg.
> 
> - Linting and (statically) validating queries.
> - Using code-generators like [GraphQL Codegen](https://the-guild.dev/graphql/codegen).

This should already be possible. You can use the `/specs/graphql` endpoint to generate the GraphQL schema for the role your using on your front-end ensuring that the contract you have in your graphql tooling for the project matches the actual queries you're allowed to make, or use the same endpoint to generate the admin-schema which includes everything (which would behave the same as your suggested change to have a single schema for everybody)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
